### PR TITLE
Improve examples

### PR DESCRIPTION
**Pull request by AlexanderDokuchaev**
### Changes

- Add pattern based on file extension to gitignore for examples
- Use pathlib style instead os.path
- Fix issue with creating ../nncf/~/.cache/datasets directory
- Replace tqdm to nncf.common.logging.track_progress
- Remove unused function arguments and if conditions
- Clean logs, disable progress bar, remove log of validataion
- Pass list of command arguments to subprocess instead of string
- Remove nncf dependencies from tests/cross_fw/examples
- Add `--reuse-venv` to use venv from example directory
- Add progress bar to collect_values_for_each_item_using_prepared_model

### Tests
https://github.com/openvinotoolkit/nncf/actions/runs/11900403097
nightly/job/test_examples/618
nightly/job/windows/job/test-examples/278/

### CODE REVIEW

File: examples/post_training_quantization/openvino/yolov8_quantize_with_accuracy_control/main.py [hunk 1]
```diff
@@ -165,13 +167,11 @@ def validation_ac(
                 ]
             preds = validator.postprocess(preds)
             validator.update_metrics(preds, batch)
-            counter += 1
         stats = validator.get_stats()
         if num_outputs == 1:
             stats_metrics = stats["metrics/mAP50-95(B)"]
         else:
             stats_metrics = stats["metrics/mAP50-95(M)"]
-        print(f"Validate: dataset length = {counter}, metric value = {stats_metrics:.3f}")
```
Comment (alexsu52) [inline examples/post_training_quantization/openvino/yolov8_quantize_with_accuracy_control/main.py:174]: Could you provide the output after your changes with timestamp? I would like to make sure that the interactivity of the example has not degenerated.

**Comment by MaximProshin** [comment]
@AlexanderDokuchaev , @alexsu52 , please run SDL checks before the merge (as you changed some imports and the way how 'path' is used). Also make sure it works on Windows.

File: examples/post_training_quantization/openvino/yolov8_quantize_with_accuracy_control/main.py (hunk 1 above)
Comment (AlexanderDokuchaev) [reply examples/post_training_quantization/openvino/yolov8_quantize_with_accuracy_control/main.py:174]: image
After add track in collect_values_for_each_item_using_prepared_model

**Comment by AlexanderDokuchaev** [comment]
@MaximProshin 
/sdl/job/nightly_trigger/346 - pass, exclude checkmarkx: warning about .github/..
nightly/job/windows/job/test-examples/278/

File: tests/cross_fw/shared/command.py [hunk 2]
```diff
@@ -17,7 +17,9 @@
 from pathlib import Path
 from typing import Any, Dict, List
 
-from nncf.common.utils.os import is_windows
+
+def is_windows() -> bool:
+    return "win32" in sys.platform
```
Comment (KodiaqQ) [inline tests/cross_fw/shared/command.py:22]: Why it was replaced? 
It looks like an unnecessary change since it is a copy of `nncf.common.utils.os.is_windows`.

File: tests/cross_fw/shared/helpers.py [hunk 3]
```diff
@@ -19,15 +19,20 @@
...
 
 TensorType = TypeVar("TensorType")
 
 
+def is_windows() -> bool:
+    return "win32" in sys.platform
+
+
+def is_linux() -> bool:
+    return "linux" in sys.platform
```
Comment (KodiaqQ) [inline tests/cross_fw/shared/helpers.py:33]: The same comment about just copying `nncf.common.utils.os.is_linux` and `nncf.common.utils.os.is_windows`.

File: .github/workflows/examples.yml [hunk 4]
```diff
@@ -46,7 +46,6 @@ jobs:
         run: cat /proc/cpuinfo
       - name: Install NNCF and test requirements
         run: |
-          pip install -e .
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply .github/workflows/examples.yml:49]: tests/cross_fw/examples - required to install NNCF only to detect os. 
After removing this dependences, NNCF is not require to run tests.

Updated

File: .gitignore [hunk 5]
```diff
@@ -119,23 +119,19 @@ nncf_debug/
...
-examples/post_training_quantization/torch/ssd300_vgg16/ssd300_vgg16_*
-examples/post_training_quantization/openvino/anomaly_stfpm_quantize_with_accuracy_control/stfpm_*
-examples/post_training_quantization/openvino/yolov8/yolov8n*
-examples/post_training_quantization/openvino/yolov8_quantize_with_accuracy_control/yolov8n*
-examples/**/runs/**
-examples/**/results/**
-examples/llm_compression/openvino/tiny_llama_find_hyperparams/statistics
-compressed_graph.dot
-original_graph.dot
-datasets/**
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply .gitignore:135]: `datasets` directory created by ultralitics, it's only happens if you have yolo config `.config/Ultralytics/settings.json` file that was generated some time ago or changes manually.

On current version of yolo, default path for `datasets_dir` generated as `nncf/../datasest` if run example from NNCF in case of  `.config/Ultralytics/settings.json`  is not exist.
If datasets directory created on your host you need to modify config by `yolo settings dataset_dir=new_path` or by text editor, or remove config file after it will be generated this new path.

File: examples/post_training_quantization/onnx/mobilenet_v2/main.py [hunk 6]
```diff
@@ -61,13 +62,17 @@ def validate(path_to_model: Path, validation_loader: torch.utils.data.DataLoader
...
+    command = [
+        "benchmark_app",
+        "-m", path_to_model.as_posix(),
+        "-d", "CPU",
+        "-api", "async",
+        "-t", "15",
+        "-shape", str(shape),
+    ]  # fmt: skip
+    cmd_output = subprocess.check_output(command, text=True)
+    print(*cmd_output.splitlines()[-8:], sep="\n")
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply examples/post_training_quantization/onnx/mobilenet_v2/main.py:75]: Dont see any depends print function from type of used progress bar. 
If you think that this changes is required, you are free to create PR.

File: examples/post_training_quantization/onnx/yolov8_quantize_with_accuracy_control/main.py [hunk 7]
```diff
@@ -155,7 +156,6 @@ def validation_ac(
         output_names = [output.name for output in session.get_outputs()]
         num_outputs = len(output_names)
 
-        counter = 0
         for batch_i, batch in enumerate(validation_loader):
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply examples/post_training_quantization/onnx/yolov8_quantize_with_accuracy_control/main.py:159]: This function calls in collect_values_for_each_item_using_prepared_model, that call validate_fn separatly for each item.
As result progress bar  creates N times.

File: examples/post_training_quantization/openvino/anomaly_stfpm_quantize_with_accuracy_control/main.py [hunk 8]
```diff
@@ -69,7 +68,6 @@ def validate(
 
     output = model.outputs[0]
 
-    counter = 0
     for batch in val_loader:
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply examples/post_training_quantization/openvino/anomaly_stfpm_quantize_with_accuracy_control/main.py:71]: https://github.com/openvinotoolkit/nncf/pull/3091#discussion_r1850075834

File: examples/post_training_quantization/openvino/yolov8_quantize_with_accuracy_control/main.py [hunk 9]
```diff
@@ -75,72 +76,74 @@ def print_statistics(stats: np.ndarray, total_images: int, total_objects: int) -
...
+def prepare_validation(model: YOLO, args: Any) -> Tuple[SegmentationValidator, torch.utils.data.DataLoader]:
+    validator: SegmentationValidator = model.task_map[model.task]["validator"](args=args)
+    validator.data = check_det_dataset(args.data)
+    validator.stride = 32
     validator.is_coco = True
     validator.class_map = coco80_to_coco91_class()
     validator.names = model.model.names
     validator.metrics.names = validator.names
     validator.nc = model.model.model[-1].nc
-    validator.nm = 32
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply examples/post_training_quantization/openvino/yolov8_quantize_with_accuracy_control/main.py:108]: It's unexpected attributes for class and not used anywhere

File: nncf/quantization/algorithms/accuracy_control/evaluator.py [hunk 10]
```diff
@@ -276,20 +277,20 @@ def collect_values_for_each_item_using_prepared_model(
...
+        :param indices: The zero-based indices of data items that should be selected from the dataset.
         :return: Collected values.
         """
+        total = len(indices) if indices is not None else dataset.get_length()
         if self._metric_mode:
             # Collect metrics for each item
             values_for_each_item = [
                 self._validation_fn(prepared_model.model_for_inference, [data_item])[0]
-                for data_item in dataset.get_data(indices)
+                for data_item in track(dataset.get_data(indices), total=total, description="Collecting metrics")
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply nncf/quantization/algorithms/accuracy_control/evaluator.py:288]: You can find logs in provided test jobs, all links are in PR description

File: tests/cross_fw/shared/command.py (hunk 2 above)
Comment (AlexanderDokuchaev) [reply tests/cross_fw/shared/command.py:22]: To remove dependencies (for only utils function) from NNCf, and avoid install NNCF for cross_fw/examples tests.

File: examples/post_training_quantization/tensorflow/mobilenet_v2/main.py
[the remark this answers is not in the document]
Comment (KodiaqQ) [reply examples/post_training_quantization/tensorflow/mobilenet_v2/main.py:1]: https://github.com/openvinotoolkit/nncf/blob/develop/examples/post_training_quantization/tensorflow/mobilenet_v2/main.py#L35

File: nncf/quantization/algorithms/accuracy_control/evaluator.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (KodiaqQ) [reply nncf/quantization/algorithms/accuracy_control/evaluator.py:288]: Please show the logs from the manual local run here. Test logs are very huge and not human-friendly.

File: examples/post_training_quantization/onnx/mobilenet_v2/main.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (KodiaqQ) [reply examples/post_training_quantization/onnx/mobilenet_v2/main.py:75]: I do not agree. Let's align it and use here and below `nncf_logger` instead of `print`s as you did for `track` instead of `tqdm`.

File: .gitignore (hunk 5 above)
[the remark this answers is not in the document]
Comment (KodiaqQ) [reply .gitignore:135]: I do not agree.
I've run the following command from the `nncf` root (as usual, for testing purposes):
`pytest -s tests/cross_fw/examples/test_examples.py -k "post_training_quantization_openvino_yolo8_quantize"`
And after the passing, I observe the following directory:
image
Thus, removing `datasets/**` from the `.gitignore` might cause issues.

[the remark this answers is not in the document]
Comment (KodiaqQ) [reply .gitignore:135]: From a high perspective, it would be more helpful if you would update examples to create a `datasets` folder during the run in a corresponding or temporary directory, to exclude incidents in the future. In this case, we would be able to remove this path from `.gitignore` and safely run tests.

[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply .gitignore:135]: Have you updated the ultralytics config file as described above?

File: examples/post_training_quantization/onnx/mobilenet_v2/main.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply examples/post_training_quantization/onnx/mobilenet_v2/main.py:75]: What do you disagree with?

[the remark this answers is not in the document]
Comment (KodiaqQ) [reply examples/post_training_quantization/onnx/mobilenet_v2/main.py:75]: I disagree with this:

> Dont see any depends print function from type of used progress bar.

I see that we should be consistent. If you're adding a progress bar from the NNCF logging part, thus add please a logging from NNCF instead of `print`s.

[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply examples/post_training_quantization/onnx/mobilenet_v2/main.py:75]: As i said before, if you are sure that using nncf_logger can improve examples for any way, you are free to create your own PR.

But i sure that nncf_logger is not should be used in examples, it's only increase output without adding any useful information for user.

@alexsu52 what do you think about it?

File: examples/post_training_quantization/tensorflow/mobilenet_v2/main.py
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply examples/post_training_quantization/tensorflow/mobilenet_v2/main.py:1]: updated 
https://github.com/openvinotoolkit/nncf/actions/runs/11934670278

File: examples/post_training_quantization/onnx/mobilenet_v2/main.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (KodiaqQ) [reply examples/post_training_quantization/onnx/mobilenet_v2/main.py:75]: @AlexanderDokuchaev, sorry, but your PR was stated as `Improve examples`. But proposed changes introduce more mess than an improvement in terms of the NNCF API usage. I don't see any reason to replace `tqdm` with the `track` then.
Please, revert the changes or align API usage everywhere.

**Review by KodiaqQ — CHANGES_REQUESTED** [review]

File: examples/post_training_quantization/tensorflow/mobilenet_v2/main.py
[the remark this answers is not in the document]
Comment (KodiaqQ) [reply examples/post_training_quantization/tensorflow/mobilenet_v2/main.py:1]: image
I do not see any updates.

File: nncf/quantization/algorithms/accuracy_control/evaluator.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply nncf/quantization/algorithms/accuracy_control/evaluator.py:288]: image

File: examples/post_training_quantization/tensorflow/mobilenet_v2/main.py
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply examples/post_training_quantization/tensorflow/mobilenet_v2/main.py:1]: sorry, missed that push failed. 
https://github.com/openvinotoolkit/nncf/pull/3091/commits/37d34c2d027c979364c36e68d3e9223f5456935d

File: .gitignore (hunk 5 above)
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply .gitignore:135]: You have config file `~/.config/Ultralytics/settings.json` on your host that was generated when you run any code with ultralitics for the first time and creating or removing any virtual environment does not affect on the config in any way.

Your config collect line like `"datasets_dir": "/home/<username>/nncf/datasets",`.
This line only generated if run NNCF examples with old version of ultralitics, current default value is like `/home/<username>/datasets` - `GIT_ROOT.parent / "datasets"`.

There is only one way conditions to get `datasets_dir` like `.../nncf/datasets`:
- You did not run any code with ultralitics before
- And for the first time you have run NNCF yolo example with old version of ultralitics

In all other cases `datasets_dir` will not set as `nncf/datasets`

My point, If you have config file of any third party package with not default values in your `~/.config` directory, that used NNCF folder as storage for anything, it's not responsibility of NNCF, it's incorrect configuration that should be resolved by user.

**Moreover**, the `datasets` directory in the NNCF git root folder breaks `isort`, by changing the import order of `datasets` module. 
So `datasets` in `.gitignore` hides this problem by making the existence of this directory expected but it is not.

@alexsu52 as you have added `datasets` to `.gitignore`, i think last word is your

[the remark this answers is not in the document]
Comment (KodiaqQ) [reply .gitignore:135]: > it's not responsibility of NNCF, it's incorrect configuration that should be resolved by user.

If I run NNCF tests from the root and NNCF creates a `datasets` folder in the root, this is NNCF's responsibility, not the user, since it is not possible to set up the folder for this via tests.
As you said, 

> datasets directory in the NNCF git root folder breaks isort

So, this is another reason to change the placing of the `datasets` folder on the NNCF's side, since tests were designed by the NNCF team.

File: examples/post_training_quantization/onnx/mobilenet_v2/main.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply examples/post_training_quantization/onnx/mobilenet_v2/main.py:75]: > Please, revert the changes or align API usage everywhere.

No, I do not agree. 

Please be consistent in your discussion, your are start this thread about replacing `print` to `nncf_logger`, Now you ask to revert change by reason that you dont see any reasons in replacing tqdm to track.

If you ask me about why `track` is more preferable then `tqdm`, i can answer:
1. Output for `tqdm` in CI or if save to file, looks too massive because each update of progress bar generate new line.
2. For tqdm can be disable but in this case output does not contain progress bar, that contain number of iterations and time, but it can be used to debug and detect issues.
3. Track based on rich, that resolve this issue and saves progress bar in one line.

To see it, you can reproduce output localy, 

```python
# script.py
from tqdm import tqdm
from time import sleep
from nncf.common.logging.track_progress import track

l = [1,2,3,4,5]
print("TQDM:")
for _ in tqdm(l):
    sleep(.5)
print("TRACK:")
for _ in track(l):
    sleep(.5)
```

run command:
```bash
python script.py  > log 2>&1
```
Output: 
```
INFO:nncf:NNCF initialized successfully. Supported frameworks detected: torch, onnx, openvino
TQDM:

  0%|          | 0/5 [00:00<?, ?it/s]
 20%|██        | 1/5 [00:00<00:02,  2.00it/s]
 40%|████      | 2/5 [00:01<00:01,  2.00it/s]
 60%|██████    | 3/5 [00:01<00:01,  2.00it/s]
 80%|████████  | 4/5 [00:02<00:00,  2.00it/s]
100%|██████████| 5/5 [00:02<00:00,  2.00it/s]
100%|██████████| 5/5 [00:02<00:00,  2.00it/s]
TRACK:
Working... ━━━━━━━━━━━━━━━━━━ 100% 5/5 • 0:00:02 • 0:00:00
```

**Comment by KodiaqQ** [comment]
@AlexanderDokuchaev, feel free to merge after approval from the other reviewers.

File: examples/post_training_quantization/onnx/mobilenet_v2/main.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (MaximProshin) [reply examples/post_training_quantization/onnx/mobilenet_v2/main.py:75]: In general, we should not overload NNCF examples with additional NNCF functionality like logging as we want users to focus on concrete API which is demonstrated in the example. So I don't think we should replace 'print's by NNCF logging for "the consistency". Regarding 'track', I agree it reduces the output but this is again NNCF functionality which doesn't belong to API that we want to demonstrate here and I would try to avoid it. @AlexanderDokuchaev , is there a way to use 'rich' directly here?
cc @alexsu52

**Review by kshpv — APPROVED** [review]

File: examples/post_training_quantization/onnx/mobilenet_v2/main.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply examples/post_training_quantization/onnx/mobilenet_v2/main.py:75]: It is possible, but default rich progress bar does not contain number of iteration and show only left rime,
and combination of two style of progress bar in one script looks weird 

default rich:
image
nncf track:
image

[the remark this answers is not in the document]
Comment (alexsu52) [reply examples/post_training_quantization/onnx/mobilenet_v2/main.py:75]: I would suggest to use the default rich.

File: .gitignore (hunk 5 above)
[the remark this answers is not in the document]
Comment (alexsu52) [reply .gitignore:135]: I read your discussion and did the following experiment:
- remove `~/.config/Ultralytics/settings.json`
- create a clear env using
```
pip install -r examples/post_training_quantization/onnx/yolov8_quantize_with_accuracy_control/requirements.txt
```

Ultralytics version:
image

- run example
```
python examples/post_training_quantization/onnx/yolov8_quantize_with_accuracy_control/main.py
```
dataset was downloaded into `nncf/datasets`
image

Where did I go wrong?

File: examples/post_training_quantization/onnx/mobilenet_v2/main.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply examples/post_training_quantization/onnx/mobilenet_v2/main.py:75]: https://github.com/openvinotoolkit/nncf/pull/3091/commits/23d8ba342fa9ade9bfedb6918f9678a7bdccb6f5
https://github.com/openvinotoolkit/nncf/actions/runs/11975030346

**Review by alexsu52 — APPROVED** [review]

**Review by MaximProshin — APPROVED** [review]

**Review by daniil-lyakhov — APPROVED** [review]
